File: src/main/java/aurocosh/divinefavor/common/lib/wrapper/PredicateWrapper.java
package aurocosh.divinefavor.common.lib.wrapper;

import java.util.function.Function;
import java.util.function.Predicate;

public class PredicateWrapper<T, K> implements Predicate<T> {
    private final Function<T, K> converter;
    private final Predicate<K> predicate;

    public PredicateWrapper(Function<T, K> converter, Predicate<K> predicate) {
        this.converter = converter;
        this.predicate = predicate;
    }

    public boolean test(T value) {
        K converted = converter.apply(value);
        return predicate.test(converted);
    }
}

File: src/main/java/aurocosh/divinefavor/common/lib/wrapper/PredicateWrapper.kt
package aurocosh.divinefavor.common.lib.wrapper

open class PredicateWrapper<T, K>(private val converter: (T) -> K, private val predicate: (K) -> Boolean) {

    fun test(value: T): Boolean {
        val converted = converter.invoke(value)
        return predicate.invoke(converted)
    }
}
